chart/templates/connection-secret-job.yaml

```diff
@@ -0,0 +1,96 @@
+{{- $db := index .Values "timescaledb-single" -}}
+{{- if and $db.enabled (eq .Values.promscale.connectionSecretName "") (eq .Values.promscale.connection.password "") -}}
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ .Release.Name }}-promscale-initializer-sa
+  namespace: {{ template "tobs.namespace" . }}
+  labels:
+    app: {{ template "tobs.fullname" . }}
+    chart: {{ template "tobs.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: {{ .Release.Name }}-promscale-initializer-role
+  namespace: {{ template "tobs.namespace" . }}
+  labels:
+    app: {{ template "tobs.fullname" . }}
+    chart: {{ template "tobs.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+rules:
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["get", "patch"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: {{ .Release.Name }}-promscale-initializer-rolebinding
+  namespace: {{ template "tobs.namespace" . }}
+  labels:
+    app: {{ template "tobs.fullname" . }}
+    chart: {{ template "tobs.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: {{ .Release.Name }}-promscale-initializer-role
+subjects:
+- namespace: {{ template "tobs.namespace" . }}
+  kind: ServiceAccount
+  name: {{ .Release.Name }}-promscale-initializer-sa
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-connection-initializer
+  namespace: {{ template "tobs.namespace" . }}
+  labels:
+    app: {{ template "tobs.fullname" . }}
+    chart: {{ template "tobs.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+data:
+  password-initializer.sh: |
+    #!/bin/bash
+    while ! kubectl get secret {{ .Release.Name }}-credentials --namespace {{ template "tobs.namespace" . }}; do
```

Comment:
we should have a maximum number of retries here

Reply:
Yeah, I was thinking if it is needed or not. Without such logic pod would end up in Pod creating state constantly and would wait until secret is present. With such logic pod would enter frequent CrashLoop and kubernetes will reschedule it indefinitely. Since the latter mean that in general there is no maximum retries (kubernetes will reschedule indefinitely) I opted for the approach without max retries. This allows to reduce pressure on kubernetes API server while keeping code simpler.

Reply:
Doesn't crash loop reschedule with exponential backoff though? In which case I think that's what we would want? Is it possible to get the logs of a pod in creating state?

Also, in terms of metrics a crash loop gives number of retries, generates k8s events and similar info no? Wouldn't that make this problem more visible and easier to debug?

I am not positive what the right thing is here but my gut is that being loud is gonna make it easier for users to debug

Reply:
> Doesn't crash loop reschedule with exponential backoff though? In which case I think that's what we would want?

I don't think exponential backoff gives us many benefits here as we are doing a simple `get` every 1s (read operations are using kubernetes API server cache, writes do not do this). It is more important to have backoff when one is creating resources to reduce the load on the database, hence this is why backoff is implemented on crash looping pods.

> Is it possible to get the logs of a pod in creating state?

Yes. And bear in mind that with crashlooping pod, logs are truncated and only part of it is available.

> Also, in terms of metrics a crash loop gives number of retries, generates k8s events and similar info no? Wouldn't that make this problem more visible and easier to debug?

Yes, with crashloop we would get more events and it will show a number of pod crashloops (not `kubectl get` retries though!). However, those events do not say anything on "why?" this pod is crashlooping and the user will still need to go and retrieve pod logs to get more information. Which gets us to the same state as with the currently implemented method. So IMHO debuggability is at the same level.

> I am not positive what the right thing is here but my gut is that being loud is gonna make it easier for users to debug

My personal take on this is that neither is loud. Without monitoring users are still expected to run `kubectl get pods` and see if pods are running or not and any state apart from `Running` is a cause for investigation. And with monitoring, we have alerts for both cases. Hence I opted for code simplicity as now it is fairly easy to understand what is happening in the script. However not to dwell on this anymore I amended the script to timeout after 10m of runtime. Such an approach shouldn't complicate code much, will allow crashing hard (crashloop), and should join both our approaches described in this thread.

Reply:
Just throwing out a random idea I had, not sure if its worthwhile but it seems like it could benefit from a timeout. I guess you could use the watch flag and check the existence of secret in the output and exiting on found.

Its a little more complex but does remove the need to run multiple `kubectl get` commands.